src/general/IWorldEffect.cs

```diff
@@ -34,71 +36,177 @@ public void OnRegisterToWorld()
+                                    totalAdded[output.Key.ID] += add;
+                                }
+                                else
+                                {
+                                    totalAdded.Add(output.Key.ID, add);
+                                }
+                            }
+                        }
+                    }
+                }
             }
 
-            totalAmount += glucoseValue.Amount;
-            initialTotalDensity += glucoseValue.Density;
-            totalChunkAmount += patch.GetTotalChunkCompoundAmount(Compound.Glucose);
-
-            // If there are microbes to be eating up the primordial soup, reduce the milk
-            if (patch.SpeciesInPatch.Count > 0)
+            // Apply results
+            foreach (var compound in patch.Biome.ChangeableCompounds)
             {
-                var initialGlucose = Math.Round(glucoseValue.Density * glucoseValue.Amount +
-                    patch.GetTotalChunkCompoundAmount(Compound.Glucose), 3);
-
-                glucoseValue.Density = Math.Max(glucoseValue.Density * targetWorld.WorldSettings.GlucoseDecay,
-                    Constants.GLUCOSE_MIN);
-
-                patch.Biome.ModifyLongTermCondition(Compound.Glucose, glucoseValue);
-
-                var finalGlucose = Math.Round(glucoseValue.Density * glucoseValue.Amount +
-                    patch.GetTotalChunkCompoundAmount(Compound.Glucose), 3);
-
-                var localReduction = Math.Round((initialGlucose - finalGlucose) / initialGlucose * 100, 1);
-
-                patch.LogEvent(new LocalizedString("COMPOUND_CONCENTRATIONS_DECREASED",
-                        glucoseDefinition.Name, new LocalizedString("PERCENTAGE_VALUE", localReduction)), false,
-                    "glucoseDown.png");
+                var tweakedBiomeConditions = compound.Value;
+
+                if (totalAdded.ContainsKey(compound.Key))
+                {
+                    if (SimulationParameters.Instance.GetCompoundDefinition(compound.Key).IsEnvironmental)
+                    {
+                        tweakedBiomeConditions.Ambient = Math.Clamp(
+                            patch.BiomeTemplate.Conditions.ChangeableCompounds[compound.Key].Ambient
+                            + totalAdded[compound.Key], 0, 1);
+                    }
+                    else
+                    {
+                        tweakedBiomeConditions.Density = Math.Clamp(
+                            patch.BiomeTemplate.Conditions.ChangeableCompounds[compound.Key].Density
+                            + totalAdded[compound.Key], 0, 1);
+                    }
+                }
+
+                targetWorld.Map.Patches[patchKeyValue.Key].Biome.ModifyLongTermCondition(compound.Key,
+                    tweakedBiomeConditions);
             }
-
-            finalTotalDensity += patch.Biome.ChangeableCompounds[Compound.Glucose].Density;
         }
+    }
+
+    private void HandlePatchCompoundDiffusion()
```

Comment:
Also wouldn't it be possible to split the compound diffusion effect to its own class? That way the separate "steps" of this change can be viewed in isolation and can be reasoned and read much more easily by programmers.